### ISSUE TITLE
Code seems desperately un-modular and hard to reuse.

### ISSUE THREAD

**Issue by mutantbob** [KyleMayes/vulkanalia#245]
The AppData class is just an ever-growing pile of related fields.  A great many of those fields should be grouped together in independent pluggable/reusable structs.

A great many functions accept an AppData just so they can access the device and queue.  Those fields could probably be grouped in a context struct that is not tangled with the buffers.

So many functions write to fields in AppData. E.g. create_command_pool() when they should probably just return the command pool and let App::create() store it in the correct field.

The App::create() function instantiates an incomplete AppData object and then calls multiple methods to mutate it.  This is not optimal code style since the data variable spends a great deal of time in an incomplete state.  Ideally there should be an AppData::new(...) that accepts the instance, device, etc.; constructs the necessary fields; and returns a valid and complete AppData object.
Basically, any place you pass &mut should be reviewed to see if it can be improved.

Solving this issue will be quite traumatic.  The code right now is optimized for brevity and simplicity of the tutorial.  "fixing" the code to be more Rust-y would be a major effort (having to redo all of the stages, both code and text narrative).

**Comment by KyleMayes** [comment]
I didn't know Vulkan when I started this project, the tutorial code is mostly a pretty direct translation of the code from the original C++ tutorial.

As you've said, significantly changing it would be a massive amount of work and I really don't have any interest in making these changes myself or even reviewing someone else'.

Maybe it would be enough to call out how janky this code is and how it mostly serves to illustrate how Vulkan works and now how one should go about structuring a Vulkan application.

**Comment by chuigda** [comment]
Recently I'm trying to create a sowewhat "bigger" Vulkan project (well though in Java instead of Rust but I think that may be useful), and I'm trying to implement a slightly better code style. You can find the relevant code [here](https://github.com/chuigda/Project-WGX/tree/master/src/main/java/chr/wgx/render/vk). I'm also a Vulkan beginner and maybe what I did is not the best practise, but hope that being useful.

**Comment by KyleMayes** [comment]
I've continued thinking about this occasionally, and I've poked around with the idea of a new tutorial (I think salvaging the existing one is probably a non-starter, though it of course contains a lot of useful material for a new tutorial).

I'm still a neophyte at building actual Vulkan applications (I admit it, my playing around with building a 2D game was done with `wgpu` instead of raw `vulkanalia`, I am a hypocrite).

The [official Vulkan tutorial](https://github.com/KhronosGroup/Vulkan-Tutorial) is based off the same tutorial I based my tutorial off of, but it uses [`Vulkan-Hpp`](https://github.com/KhronosGroup/Vulkan-Hpp) which (among other things) provides a RAII wrapper around Vulkan resources. I haven't gone through the official tutorial in-depth yet, but it doesn't seem a whole lot better in terms of code organization other than not needing to explicitly destroy stuff (which is indeed good).

I think there is room for a new Vulkan + Rust tutorial, one that builds up from the thin Vulkan wrapper offered by `vulkanalia` into a simple framework that uses a modern approach to Vulkan usage and is more Rusty and flexible than dumping every type of Vulkan handle imaginable into a single struct.

But... I'm not sure I'm qualified to provide something like that. So I'm not sure why I am posting this comment, but I have been thinking about this more and more and come the new year I think I will start spending some more time and effort on this.

So if you have some ideas for what to do, please feel free to post them here.
